### PR TITLE
configure the ingress controller for k8s deployed jobs

### PR DESCRIPTION
**Pull request by ebiiii**

### CODE REVIEW

File: renga_deployer/engines.py [hunk 1]
```diff
@@ -228,6 +232,10 @@ def launch(self, execution, engine=None, **kwargs):
         """Launch a Kubernetes Job with the context spec."""
         context = execution.context
 
+        execution.environment.update({
+            'DEPLOYER_BASE_URL': '/interactive/{0}'.format(execution.id)
```
Comment (rokroskar) [inline renga_deployer/engines.py]: where is this used?

File: renga_deployer/engines.py [hunk 2]
```diff
@@ -498,20 +522,47 @@ def get_host_ports(self, execution):
...
-                        'DEPLOYER_K8S_CONTAINER_IP'] or pod.items[
-                            0].status.host_ip,
-                    'exposed':
-                    port.node_port,
-                    'protocol':
-                    port.protocol,
-                } for port in service.items[0].spec.ports]
-            }
+            if current_app.config.get(
+                    'DEPLOYER_K8S_USE_INGRESS'):   # pragma no cover
```
Comment (jirikuncar) [inline renga_deployer/engines.py]: Can we add tests instead ignoring the code?

--
https://medium.com/@Oskarr3/setting-up-ingress-on-minikube-6ae825e98f82

File: renga_deployer/engines.py [hunk 3]
```diff
@@ -498,20 +522,47 @@ def get_host_ports(self, execution):
...
+                        'path':
+                        ingress.spec.rules[0].http.paths[0].path,
+                        'exposed':
+                        '443',
+                        'protocol':
+                        port.protocol,
+                    } for port in service.items[0].spec.ports]
+                }
+
+            else:
```
Comment (jirikuncar) [inline renga_deployer/engines.py]: Using `else` & `return` is an anti-pattern in Python. Can you remove the else and remove one level of indentation?

File: renga_deployer/engines.py (hunk 1 above)
Comment (ebiiii) [reply renga_deployer/engines.py]: https://github.com/SwissDataScienceCenter/renga-ui/blob/deploy_kube/src/utils/renga-api.ts#L75
It is used in the command of the container, to make it aware of the prefix of its path, dynamically.

File: renga_deployer/engines.py (hunk 2 above)
Comment (ebiiii) [reply renga_deployer/engines.py]: I added tests covering all the ingress parts

File: tests/test_deployer.py [hunk 4]
```diff
@@ -85,15 +86,69 @@ def test_open_port(app, engine, image, deployer):
...
+    try:
+        binding = deployer.get_host_ports(execution)['ports'][0]
+        s.connect((binding['host'], int(binding['exposed'])))
+        phrase = b'earth_calling'
+        s.send(phrase)
+        received = s.recv(100)
+        assert received == phrase
+        assert execution.check_state(ExecutionStates.RUNNING,
+                                     deployer.ENGINES[engine]())
+    except Exception as e:
```
Comment (jirikuncar) [inline tests/test_deployer.py]: IMHO you can skip this section

File: renga_deployer/engines.py (hunk 1 above)
Comment (rokroskar) [reply renga_deployer/engines.py]: it should be documented with a default provided in `config.py`

File: tests/test_deployer.py (hunk 4 above)
Comment (rokroskar) [reply tests/test_deployer.py]: @jirikuncar: it's useful to quickly check that the connections to the containers are being made correctly

**Review by rokroskar — DISMISSED** [review]

File: renga_deployer/config.py [hunk 5]
```diff
@@ -65,6 +65,14 @@
 DEPLOYER_K8S_USE_INGRESS = False
 """Whether to use an ingress for reaching k8s endpoints."""
 
+DEPLOYER_K8S_INGRESS_CLASS = 'nginx'
```
Comment (jirikuncar) [inline renga_deployer/config.py]: From practical point of view, can we join `DEPLOYER_K8S_USE_INGRESS` and `DEPLOYER_K8S_INGRESS_CLASS` configuration options?

```
DEPLOYER_K8S_INGRESS = None  # or False disables ingress (default)
DEPLOYER_K8S_INGRESS = 'nginx'  # would enable ingress with 'nginx' class
```

File: renga_deployer/engines.py [hunk 6]
```diff
@@ -96,6 +96,11 @@ def launch(self, execution, **kwargs):
         """Launch a docker container with the context image."""
         context = execution.context
 
+        execution.environment.update({
```
Comment (jirikuncar) [inline renga_deployer/engines.py]: Q: shouldn't this be `execution.environment.setdefault('DEPLOYER_BASE_URL', current_app.config.get('DEPLOYER_DEFAULT_BASE_URL'))`?

File: renga_deployer/engines.py [hunk 7]
```diff
@@ -228,6 +241,13 @@ def launch(self, execution, engine=None, **kwargs):
         """Launch a Kubernetes Job with the context spec."""
         context = execution.context
 
+        execution.environment.update({
```
Comment (jirikuncar) [inline renga_deployer/engines.py]: again, I would say we should use `setdefault` instead of `update`.

File: renga_deployer/engines.py [hunk 8]
```diff
@@ -260,12 +280,13 @@ def launch(self, execution, engine=None, **kwargs):
...
             if current_app.config.get(
-                    'DEPLOYER_K8S_USE_INGRESS'):  # pragma no cover
+                    'DEPLOYER_K8S_USE_INGRESS'):
                 beta_api = self._kubernetes.client.ExtensionsV1beta1Api()
                 ingress = beta_api.create_namespaced_ingress(
                     namespace,
                     self._k8s_ingress_template(uid, service.metadata.name,
-                                               context.spec['ports'][0]))
+                                               context.spec['ports'][0],
+                                               execution.id))
```
Comment (jirikuncar) [inline renga_deployer/engines.py]: I would propose, that we pass only the `execution` to `_k8s_ingress_template` and the `execution.id` and `execution.context` is extracted inside the method.

File: renga_deployer/engines.py [hunk 9]
```diff
@@ -440,22 +461,37 @@ def _k8s_service_template(namespace, context, uid):
 
     @staticmethod
     def _k8s_ingress_template(uid, service_name,
```
Comment (jirikuncar) [inline renga_deployer/engines.py]: Seeing changes on this method, we might consider new signature: `__k8s_ingress_template(uid, service, execution)` (`service_name = service.metadata.name; execution_id = execution.id; service_port = execution.context.ports[0]`).

File: renga_deployer/config.py (hunk 5 above)
Comment (ebiiii) [reply renga_deployer/config.py]: make sense. I'll change it.

File: renga_deployer/engines.py (hunk 6 above)
Comment (ebiiii) [reply renga_deployer/engines.py]: A: I don't have a case in mind where the variable would have been set before and could be overwritten with something different here. So both options work for me.

**Review by jirikuncar — APPROVED** [review]

**Pull request merged**